prov/efa/src/rxr/rxr_msg.c

```diff
@@ -93,10 +98,37 @@ ssize_t rxr_msg_post_rtm(struct rxr_ep *rxr_ep, struct rxr_tx_entry *tx_entry)
 		return rxr_pkt_post_ctrl_or_queue(rxr_ep, RXR_TX_ENTRY, tx_entry, rtm_type + tagged, 0);
 	}
 
+	if (delivery_complete_requested_flag && !(peer->flags & RXR_PEER_HANDSHAKE_RECEIVED)) {
+		/*
+		 * The sender cannot send with FI_DELIVERY_COMPLETE
+		 * if the peer is not able to handle it.
+		 * If the sender does not know whether the peer
+		 * can handle it, it needs to wait for
+		 * a handshake packet from the peer. The handshake
+		 * packet contains the information whether the peer
+		 * support it or not.
+		 */
+		if (!(peer->flags & RXR_PEER_HANDSHAKE_RECEIVED)) {
+			err = rxr_pkt_wait_handshake(rxr_ep, tx_entry, peer);
+			if (OFI_UNLIKELY(err))
+				return err;
+		}
+
+		if (delivery_complete_requested_flag &&
+		    peer->flags & RXR_PEER_HANDSHAKE_RECEIVED &&
```

Comment:
This piece of code is already inside ```if (delivery_complete_requested_flag)```, so the first condition is redundant.

Also, this is called after successful return of ```rxr_pkt_wait_handshake```. so ```peer->flags & RXR_PEER_HANDSHAKE_RECEIVED``` should be an assertion.